stunnel/stunnel_cache.ml

```diff
@@ -146,31 +143,26 @@ let gc () = Mutex.execute m unlocked_gc
 let counter = ref 0
 
 let add (x: Stunnel.t) =
-  if x.Stunnel.legacy && not (Stunnel.is_legacy_protocol_and_ciphersuites_allowed ()) then (
-    info "Legacy-protocol stunnel (id=%s) not allowed in cache: disconnecting." (id_of_stunnel x);
-    Stunnel.disconnect ~force:true x
-  ) else (
-    let now = Unix.gettimeofday () in
-    Mutex.execute m (fun () ->
-        let idx = !counter in
-        incr counter;
-        Hashtbl.add !times idx now;
-        Hashtbl.add !stunnels idx x;
-        let ep = { host = x.Stunnel.host; port = x.Stunnel.port; verified = x.Stunnel.verified } in
-        let existing =
-          if Hashtbl.mem !index ep
-          then Hashtbl.find !index ep
-          else [] in
-        Hashtbl.replace !index ep (idx :: existing);
-        debug "Adding stunnel id %s (idle %.2f) to the cache"
-          (id_of_stunnel x) 0.;
-        unlocked_gc ()
-      )
+  let now = Unix.gettimeofday () in
+  Mutex.execute m (fun () ->
+      let idx = !counter in
+      incr counter;
+      Hashtbl.add !times idx now;
+      Hashtbl.add !stunnels idx x;
+      let ep = { host = x.Stunnel.host; port = x.Stunnel.port; verified = x.Stunnel.verified } in
+      let existing =
```

Comment:
Use `Hashtbl.find` (and deal with `Not_found`) or `Hashtbl.find_opt` and avoid a race between `mem` and `find`.

Reply:
while I agree that it should be changed, I think it's a matter of style: this is executed while holding the mutex, otherwise we'd have races with the following `replace` as well.